### DISCUSSION TITLE
Allow to pass closure for icon() method

### DISCUSSION THREAD

**Discussion by studiowizjo** [filamentphp/filament#1474] [Help] [answered]
I have table push action to toggle status of project (archived or current). It would be nice if you can pass closure for icon method, so you can use different icons for each state.

**Comment by danharrin** [comment]
As far as I know, you can already! `->icon(fn ($record) => do something with $record and return the icon)`

**Reply by studiowizjo** [reply]
Unfortunately not. Here is my code:

```php
->pushActions([
	IconButtonAction::make('toggle_finish')
		->action(fn (Meeting $record) => $record->toggleFinish())
		->color('warning')
		->icon(fn ($record) => $record->active ? 'heroicon-o-lock-closed' : 'heroicon-o-lock-open'),
])
```

and it generates error:

```php
Filament\Tables\Actions\Action::icon(): Argument #1 ($icon) must be of type string, Closure given...
```

Definition of this method comes from there:

```php
namespace Filament\Tables\Actions\Concerns;

trait HasIcon
{
    protected ?string $icon = null;

    public function icon(string $icon): static
    {
        $this->icon = $icon;

        return $this;
    }

    public function getIcon(): ?string
    {
        return $this->icon;
    }
}
```

**Comment by danharrin** [comment, marked as answer]
Hm weird, are you using an old version? This is the latest `HasIcon` - https://github.com/laravel-filament/filament/blob/2.x/packages/tables/src/Actions/Concerns/HasIcon.php

**Reply by studiowizjo** [reply]
I was using 2.x, (2.0?) not sure which exactly. After update works perfectly, thanks =).
